### ISSUE TITLE
Global mouse entered and exited events

### ISSUE THREAD

**Issue by cagdassalur** [FelixKratz/SketchyBar#178]
What i want to achieve is to be able to hover over an item and select something from the popup menu. 
Right now when when my mouse exits the item and enters the popup, popup goes away naturally as i bound mouse.exit to close the popup.
Is there a way i can let the popup close when mouse exits the item to sides for example but not when mouse goes over the popup?
I have this for example:
```
	sketchybar -m --add item upcoming left \
		--set upcoming update_freq=20 \
		--set upcoming updates=on \
		--set upcoming popup.background.color=0xff000000 \
		--subscribe upcoming mouse.entered mouse.exited \
		--set upcoming script="python3 ~/.config/sketchybar/plugins/upcoming.py" 
```
and the script has: 
```
def hover(sender):
    if 'entered' in sender:
        os.system('sketchybar -m --set upcoming popup.drawing=on')
    elif 'exited' in sender:
        os.system('sketchybar -m --set upcoming popup.drawing=off')
```

https://user-images.githubusercontent.com/3230608/165513471-fdbea71b-7a6b-481e-9f9b-c08b03acd3a8.mov

I guess one option for me could be to have popup items bound to another script (currently click_script is an `open` command with a link for popup items) and enable drawing there as well, but this will be messy and it could cause a flicker (didn't try this).
Or i could hold a system-wide state somewhere to decide when to properly toggle the popup menu, but it'll be still ugly.

Sharing the issue in case you might see a value on having some sort of internal functionality for this.. However i don't see how this could be implemented in a nice way internally as well.  

edit: ah maybe i can use brackets with a seperate script for hover, let's see
edit2: ok i don't think brackets support either scripts or mouse events

**Comment by FelixKratz** [comment]
With the current capabilities this is not possible in a non janky way I think.

The way this could be approached internally is to create a new event that triggers when the mouse moves out of *all* sketchybar windows. However, this has the potential to be a bit confusing since there is already the `mouse.entered` and `mouse.exited` event for the specific items. The naming of the "mouse entered anywhere" and "mouse left everywhere" would need to be very precise to not lead to major confusion.

If nothing better comes to mind we could go with the two new events:
* `mouse.entered.globally`
* `mouse.exited.globally`

Then you could simply subscribe to the `mouse.exited.globally` event to close the popup.

Would this fix your problem?

**Comment by cagdassalur** [comment]
Yeah it would be quite nice to have something like this. And with `mouse.entered.globally` we can do make some items only shown in the bar when mouse is over. Similar to this [bartender](https://www.macbartender.com/) functionality: 
![quickreveal](https://user-images.githubusercontent.com/3230608/165688845-8e1f1d60-0d9d-4ed1-9a6c-ddda7f02b81e.gif)

**Comment by FelixKratz** [comment]
With #148 we can even have the item showing and hiding animated.
I will have to rework the way my tracking rects work for these new events, will look at it after #148.

**Comment by cagdassalur** [comment]
> With #148 we can even have the item showing and hiding animated. I will have to rework the way my tracking rects work for these new events, will look at it after #148.

oh nice! looking forward to it, thanks for the update!
I was thinking of maybe tackling this ticket myself but as a lowly ruby/java dev sadly i have no idea what's going on :)
Best of luck to you

**Comment by FelixKratz** [comment]
Since 20c3326e87c90a40c41d4fe609f206a065c3807d popups are respected in the global mouse exited event and the event will only trigger once both the bar and all popups are left. Additionally I have made it such that the `mouse.exited` event for individual items is not triggered if `mouse.exited.global` is also subscribed and the mouse enters over a popup.
So if you subscribe to `mouse.entered` `mouse.exited` and `mouse.exited.global` you will have the desired behavior:

https://user-images.githubusercontent.com/22680421/173207324-8c27d577-6265-467b-a638-2ec1f770b134.mp4

**Comment by cagdassalur** [comment]
Hey, i've only been able to test now, works perfectly!
Thanks a lot for this, really appreciate it ❤️
